src/Orleans/Configuration/ProviderConfiguration.cs

```diff
@@ -193,23 +193,39 @@ public IList<IProvider> Children
     [Serializable]
     public class ProviderCategoryConfiguration
     {
-        public string Name { get; set; }
+        public ProviderCategoryConfiguration()
+        {
+            Providers = new Dictionary<string, IProviderConfiguration>();
+        }
 
-        public IDictionary<string, IProviderConfiguration> Providers { get; set; }
+        internal ProviderCategoryConfiguration(string name) : this()
```

Comment:
It doesn't show here (or do I not only see), but should all the constructor parameters be checked, like on [line 57](https://github.com/yevhen/orleans/blob/programmatic_bootstrapper_config/src/Orleans/Configuration/ProviderConfiguration.cs#L57). I would look like so since one could try to pass null here, but otherwise ``Providers``, for instance, would be an empty collection, which leads to "diverging" code.